### PR TITLE
[trainer] fix bug in grad accum with multiple epochs

### LINKED ISSUE #22082: Inconsistent training steps between Trainer and DeepSpeed
### System Info

- `transformers` version: 4.26.0
- Platform: Linux-5.4.0-136-generic-x86_64-with-glibc2.17
- Python version: 3.8.16
- Huggingface_hub version: 0.12.1
- PyTorch version (GPU?): 1.12.0+cu113 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes

DeepSpeed general environment info:
torch install path ............... ['/home/fenia/anaconda3/envs/benchmark/lib/python3.8/site-packages/torch']
torch version .................... 1.12.0+cu113
deepspeed install path ........... ['/home/fenia/anaconda3/envs/benchmark/lib/python3.8/site-packages/deepspeed']
deepspeed info ................... 0.8.1, unknown, unknown
torch cuda version ............... 11.3
torch hip version ................ None
nvcc version ..................... 11.8
deepspeed wheel compiled w. ...... torch 1.12, cuda 11.3

### Who can help?

@stas00 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Hello! 

There seems to be some incosistency in the number of training steps when using DeepSpeed with HF trainer. 
It looks like DeepSpeed is doing things correctly but ends up training more steps in order to match Trainer. They both continue training even after learning rate has dropped to 0.

From the official examples:
```
ds_config_zero2={
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },

    "bf16": {
        "enabled": "auto"
    },

    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupDecayLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto",
            "total_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 2,
        "allgather_partitions": true,
        "allgather_bucket_size": 2e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 2e8,
        "contiguous_gradients": true
    },

    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "steps_per_print": 10,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false
}
```

```
DISTRIBUTED_ARGS="--nproc_per_node 2 --nnodes 1 --node_rank 0 --master_addr localhost --master_port 6000"
python -m torch.distributed.launch $DISTRIBUTED_ARGS \
	run_clm.py \
	--model_name_or_path gpt2 \
	--dataset_name wikitext \
	--dataset_config_name wikitext-103-raw-v1 \
	--per_device_train_batch_size 2 \
	--per_device_eval_batch_size 1 \
	--do_train \
	--output_dir /tmp/test-clm2 \
	--max_train_samples=148 \
	--gradient_accumulation_steps=16 \
	--overwrite_output_dir \
	--max_steps=200 \
	--logging_steps=10 \
	--deepspeed="ds_config_zero2.json"
```

I attach the training output: [output.txt](https://github.com/huggingface/transformers/files/10934213/output.txt)

The same behavior is observed even if training with Trainer+DeepSpeed on a single GPU.

### Expected behavior

Expected number of steps should match between Trainer and DeepSpeed logging.

Thank you very much in advance!

### PR DESCRIPTION
**Pull request by stas00**
Please see https://github.com/huggingface/transformers/issues/22082 for the analysis printout of the problem.

But basically we have a bug in grad accum machinery when `steps_in_epoch % gradient_accumulation_steps != 0`

we always check for `step+1 % gradient_accumulation_steps != 0` and when we hit the epoch boundary we end up running more than `gradient_accumulation_steps` in that iteration.

I proposed a fix using a total step counter - please feel free to suggest a different fix.

I left the debug prints if you'd like to validate the situation yourself. will remove when happy.

Fixes: https://github.com/huggingface/transformers/issues/22082

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1831,6 +1831,7 @@ def _inner_training_loop(
                     # AT THE VERY END!
                     _ = list(train_dataloader.sampler)
 
+        total_steps = 0
```
Comment (stas00) [inline src/transformers/trainer.py:1834]: let me know if you prefer to start with -1, so it'll be in sync with `step` which starts counting from -1.

**Review by sgugger — APPROVED** [review]
Indeed. Originally this was to always do a step at the end of the epoch and re-start in the next epoch, but then we change that functionality to only do a step at each epoch when the number of samples is smaller than the gradient accumulation size.

Thanks a lot for fixing, I just have one comment on naming but then good to go once the debugging statements are removed.

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py:1834]: The step=-1 is mostly to make sure that variable is initialized even if we don't get into the loop for an epoch, so don't worry about it. This is fine.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1867,6 +1868,7 @@ def _inner_training_loop(
 
             step = -1
             for step, inputs in enumerate(epoch_iterator):
+                total_steps += 1
```
[the remark this answers is not in the document]
Comment (stas00) [reply src/transformers/trainer.py:1871]: hmm, but it's not `total_samples` as each pull from a DL can be more than one sample if bs>1.

[the remark this answers is not in the document]
Comment (stas00) [reply src/transformers/trainer.py:1871]: also I think this bug will still be there for resumes unless the total counter is set to a saved at the end of the previous session value.

It's insignificant in the large scale, but it's likely to lead to a spike in the loss and other values on resume since the calculation of the averaged values will be wrong when divided by GAS.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/trainer.py:1871]: Or `total_batches` if you prefer.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/trainer.py:1871]: As for resuming, the resuming necessarily happens at a training step, so a round multiple of number of gradient accumulation batches. Therefore there should be no problem.

[the remark this answers is not in the document]
Comment (stas00) [reply src/transformers/trainer.py:1871]: > Or `total_batches` if you prefer.

I thought of this one, but I think it too is ambiguous due to GAS. 

May be `total_batched_samples`? That would be the closest I think if we want to speak DL and not steps.

a. `total_across_epochs_steps`
b. `total_batched_samples`
c. ???

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/trainer.py:1871]: `total_batched_samples` works for me too. I just want to avoid steps as I know many people assume it's the number of `optimizer.step()` done.

[the remark this answers is not in the document]
Comment (stas00) [reply src/transformers/trainer.py:1871]: I hear you - totally, let's go with `total_batched_samples` then.
